src/hotspot/os/aix/porting_aix.cpp

```diff
@@ -1042,38 +1044,58 @@ void* Aix_dlopen(const char* filename, int Flags, const char** error_report) {
     result = ::dlopen(filename, Flags);
     assert(result == nullptr, "dll_load: Could not stat() file %s, but dlopen() worked; Have to improve stat()", filename);
   #endif
+    *error_report = "Could not load module .\nSystem error: No such file or directory";
+    return nullptr;
   }
   else {
     int i = 0;
     TableLocker lock;
     // check if library belonging to filename is already loaded.
     // If yes use stored handle from previous ::dlopen() and increase refcount
     for (i = 0; i < g_handletable_used; i++) {
-      if (g_handletable[i].handle &&
-          g_handletable[i].inode == libstat.st_ino &&
-          g_handletable[i].devid == libstat.st_dev) {
-        g_handletable[i].refcount++;
-        result = g_handletable[i].handle;
+      if ((p_handletable + i)->handle &&
+          (p_handletable + i)->inode == libstat.st_ino &&
+          (p_handletable + i)->devid == libstat.st_dev) {
+        (p_handletable + i)->refcount++;
+        result = (p_handletable + i)->handle;
         break;
       }
     }
     if (i == g_handletable_used) {
       // library not yet loaded. Check if there is space left in array
       // to store new ::dlopen() handle
       if (g_handletable_used == max_handletable) {
-        // Array is already full. No place for new handle. Cry and give up.
-        *error_report = "dlopen: too many libraries loaded";
-        assert(false, "max_handletable reached");
-        return nullptr;
+        // No place in array anymore.
+        if (max_handletable == 0) {
+          // First time we allocate memory for 128 Entries
+          char* ptmp = (char*)::malloc(128 * sizeof(struct handletableentry));
```

Comment:
No need for malloc. You can start with realloc, since realloc(NULL, ...) is malloc.

```
static handletablentry* tab = nullptr;
static unsigned max_handles = 0;
...

if (need more handles)
   unsigned new_max = MAX2(max_handles * 2, init_num_handles);
   handleentry* new_tab = ::realloc(p_handletable, sizeof(handleentry) * new_max);
   if (new_tab != nullptr) {
     max_handles = new_max;
     tab= new_tab;
   }
```